pkg/agent/proxy/proxier.go

```diff
@@ -141,29 +141,30 @@ func (p *proxier) isInitialized() bool {
 	return p.endpointsChanges.Synced() && p.serviceChanges.Synced()
 }
 
-// removeStaleServices removes all expired Services. Once a Service is deleted, all
-// its Endpoints will be expired, and the removeStaleEndpoints method takes
-// responsibility for cleaning up, thus we don't need to call removeEndpoint in this
-// function.
+// removeStaleServices removes all the configurations of expired Services and their associated Endpoints.
 func (p *proxier) removeStaleServices() {
 	for svcPortName, svcPort := range p.serviceInstalledMap {
 		if _, ok := p.serviceMap[svcPortName]; ok {
 			continue
 		}
 		svcInfo := svcPort.(*types.ServiceInfo)
-		klog.V(2).Infof("Removing stale Service: %s %s", svcPortName.Name, svcInfo.String())
+		klog.V(2).InfoS("Removing stale Service", "Service", svcPortName)
 		if err := p.ofClient.UninstallServiceFlows(svcInfo.ClusterIP(), uint16(svcInfo.Port()), svcInfo.OFProtocol); err != nil {
 			klog.ErrorS(err, "Failed to remove flows of Service", "Service", svcPortName)
 			continue
 		}
-
-		if p.proxyAll {
-			// Remove NodePort flows and configurations.
-			if svcInfo.NodePort() > 0 {
-				if err := p.uninstallNodePortService(uint16(svcInfo.NodePort()), svcInfo.OFProtocol); err != nil {
-					klog.ErrorS(err, "Failed to remove flows and configurations of Service", "Service", svcPortName)
-					continue
-				}
+		// Remove associated Endpoints flows.
+		if endpoints, ok := p.endpointsInstalledMap[svcPortName]; ok {
+			if err := p.removeEndpoints(endpoints, svcInfo.Protocol()); err != nil {
+				klog.ErrorS(err, "Failed to remove flows of Endpoints for Service", "Service", svcPortName)
+				continue
+			}
+		}
+		// Remove NodePort flows and configurations.
+		if p.proxyAll && svcInfo.NodePort() > 0 {
+			if err := p.uninstallNodePortService(uint16(svcInfo.NodePort()), svcInfo.OFProtocol); err != nil {
+				klog.ErrorS(err, "Failed to remove flows and configurations of Service", "Service", svcPortName)
```

Comment:
this error log sounds quite generic, given that this case is specific to NodePort Services?